runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
+     * @param wires Qubits to apply the operation to.
+     * @param inverse Apply the inverse (Hermitian adjoint) of the operation.
+     * @param controlled_wires Control qubits applied to the operation.
+     * @param controlled_values Control values associated to the control qubits (equal length).
+     */
+    virtual void NamedOperation(const std::string &name, const std::vector<double> &params,
+                                const std::vector<QubitIdType> &wires, bool inverse = false,
+                                const std::vector<QubitIdType> &controlled_wires = {},
+                                const std::vector<bool> &controlled_values = {}) = 0;
     /**
-     * @brief Start recording a quantum tape if provided.
+     * @brief Perform a computational-basis measurement on one qubit.
+     *
+     * This instruction is generally used for mid-circuit measurements, implementing an immediate,
+     * random projective measurement and producing a classical result as output. However, it may
+     * also be used in terminal measurements where the measurement processes below are unsupported.
      *
-     * @note This is backed by the `Catalyst::Runtime::CacheManager<ComplexT>` property in
-     * the device implementation.
+     * @param wire The qubit to measure.
+     * @param postselect Optional parameter to force the result to the provided state (roughly
+     *                   equivalent to post-selection).
+     *
+     * @return `Result` The measurement result.
      */
-    virtual void StartTapeRecording() = 0;
+    virtual auto Measure(QubitIdType wire, std::optional<int32_t> postselect) -> Result = 0;
 
     /**
-     * @brief Stop recording a quantum tape if provided.
+     * @brief (Optional) Apply an arbitrary Unitary matrix to the device.
      *
-     * @note This is backed by the `Catalyst::Runtime::CacheManager<ComplexT>` property in
-     * the device implementation.
+     * Instead of identifying an operation by name, this instruction uses the mathematical
+     * representation of a quantum operator as a Unitary matrix in the computational basis.
+     *
+     * See `NamedOperation` for additional gate semantics.
+     *
+     * @param matrix A 1D array representation of the matrix in row-major format.
+     * @param wires Qubits to apply the operation to.
+     * @param inverse Apply the inverse of the operation.
+     * @param controlled_wires Control qubits applied to the operation.
+     * @param controlled_values Control values associated to the control qubits (equal length).
      */
-    virtual void StopTapeRecording() = 0;
+    virtual void
+    MatrixOperation([[maybe_unused]] const std::vector<std::complex<double>> &matrix,
+                    [[maybe_unused]] const std::vector<QubitIdType> &wires, bool inverse = false,
+                    [[maybe_unused]] const std::vector<QubitIdType> &controlled_wires = {},
+                    [[maybe_unused]] const std::vector<bool> &controlled_values = {})
```

Comment:
Same here... 
```suggestion
                    [[maybe_unused]] const std::vector<QubitIdType> &wires, bool inverse,
                    [[maybe_unused]] const std::vector<QubitIdType> &controlled_wires,
                    [[maybe_unused]] const std::vector<bool> &controlled_values)
```

Reply:
Adding it to the TODOs, but I'm mainly interested in documenting the semantics well.